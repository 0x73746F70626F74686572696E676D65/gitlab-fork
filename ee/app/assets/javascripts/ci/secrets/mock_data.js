const mockSecretsData = [
  {
    name: 'runner-token-windows',
    labels: [],
    lastAccessed: '2024-03-19T20:55:08.551Z',
    createdAt: '2024-03-11T01:33:06.258Z',
  },
  {
    name: 'product/client-tokens',
    labels: ['env::staging', 'env::production'],
    lastAccessed: '2024-01-23T07:12:44.833Z',
    createdAt: '2024-01-22T08:04:26.024Z',
  },
  {
    name: 'product/server-tokens',
    labels: ['env::staging', 'env::production'],
    lastAccessed: '2024-01-22T07:12:44.833Z',
    createdAt: '2023-01-24T08:04:26.024Z',
  },
  {
    name: 'security-token-project2',
    labels: ['env::production', 'security'],
    lastAccessed: '2024-01-21T23:13:37.227Z',
    createdAt: '2024-01-19T11:59:43.516Z',
  },
  {
    name: 'personal-tokens',
    labels: [],
    lastAccessed: '2024-01-19T21:01:38.925Z',
    createdAt: '2024-01-17T05:51:54.602Z',
  },
  {
    name: 'token-deployment',
    labels: [],
    lastAccessed: '2024-01-16T23:13:31.548Z',
    createdAt: '2024-01-15T17:12:33.882Z',
  },
  {
    name: 'runner-token-mac',
    labels: [],
    lastAccessed: '2024-01-06T07:50:58.308Z',
    createdAt: '2023-01-15T16:35:38.326Z',
  },
  {
    name: 'aws-db-prod-credentials',
    labels: ['AWS'],
    lastAccessed: '2024-01-05T11:47:59.454Z',
    createdAt: '2024-01-04T10:37:41.135Z',
  },
  {
    name: 'runner-token-linux',
    labels: [],
    lastAccessed: '2023-12-30T16:41:08.551Z',
    createdAt: '2023-01-14T18:52:56.258Z',
  },
  {
    name: 'SSH-KEY-1',
    labels: [],
    lastAccessed: '2023-11-29T22:18:31.562Z',
    createdAt: '2023-11-30T01:29:14.394Z',
  },
  {
    name: 'group-ID',
    labels: ['shell'],
    lastAccessed: '2023-11-29T04:30:10.220Z',
    createdAt: '2023-01-21T17:34:33.544Z',
  },
  {
    name: 'security-token-project1',
    labels: ['security'],
    lastAccessed: '2023-11-28T04:30:10.220Z',
    createdAt: '2023-01-21T17:34:33.544Z',
  },
];

export const mockGroupSecretsData = mockSecretsData.map((secret, index) => ({
  key: `group_secret_${index}`,
  ...secret,
}));

export const mockProjectSecretsData = mockSecretsData.map((secret, index) => ({
  key: `project_secret_${index}`,
  ...secret,
}));
