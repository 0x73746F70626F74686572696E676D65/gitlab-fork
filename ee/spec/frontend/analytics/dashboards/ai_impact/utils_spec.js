import {
  generateDateRanges,
  generateTableColumns,
  generateSkeletonTableData,
  generateTableRows,
  calculateCodeSuggestionsUsageRate,
} from 'ee/analytics/dashboards/ai_impact/utils';
import { mockTimePeriods } from './mock_data';

describe('AI impact Dashboard utils', () => {
  describe('generateDateRanges', () => {
    it.each`
      date            | description
      ${'07-01-2021'} | ${'on the first of the month'}
      ${'03-31-2021'} | ${'on the last of the month'}
      ${'03-31-2020'} | ${'in a leap year'}
    `('returns the expected date ranges $description', ({ date }) => {
      expect(generateDateRanges(new Date(date))).toMatchSnapshot();
    });
  });

  describe('generateTableColumns', () => {
    it.each`
      date            | description
      ${'07-01-2021'} | ${'on the first of the month'}
      ${'03-31-2021'} | ${'on the last of the month'}
      ${'03-31-2020'} | ${'in a leap year'}
    `('returns the expected table fields $description', ({ date }) => {
      expect(generateTableColumns(new Date(date))).toMatchSnapshot();
    });
  });

  describe('generateSkeletonTableData', () => {
    it('returns the skeleton based on the table fields', () => {
      expect(generateSkeletonTableData()).toMatchSnapshot();
    });
  });

  describe('generateTableRows', () => {
    it('returns the data formatted as a table row', () => {
      expect(generateTableRows(mockTimePeriods)).toMatchSnapshot();
    });
  });

  describe('calculateCodeSuggestionsUsageRate', () => {
    it('returns null when counts are undefined', () => {
      expect(calculateCodeSuggestionsUsageRate()).toBeNull();
    });

    it('returns null when there is no code suggestions usage data', () => {
      expect(
        calculateCodeSuggestionsUsageRate({
          codeSuggestionsContributorsCount: 0,
          codeContributorsCount: 0,
        }),
      ).toBeNull();
    });

    it('returns the code suggestions usage rate as expected', () => {
      expect(
        calculateCodeSuggestionsUsageRate({
          codeSuggestionsContributorsCount: 3,
          codeContributorsCount: 4,
        }),
      ).toEqual(75);
    });
  });
});
