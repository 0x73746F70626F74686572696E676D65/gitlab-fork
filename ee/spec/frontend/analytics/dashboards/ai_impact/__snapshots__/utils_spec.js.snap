// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AI impact Dashboard utils generateDateRanges returns the expected date ranges in a leap year 1`] = `
Array [
  Object {
    "end": 2019-10-31T23:59:59.000Z,
    "key": "5-months-ago",
    "label": "Oct",
    "start": 2019-10-01T00:00:00.000Z,
    "thClass": "gl-w-10p",
  },
  Object {
    "end": 2019-11-30T23:59:59.000Z,
    "key": "4-months-ago",
    "label": "Nov",
    "start": 2019-11-01T00:00:00.000Z,
    "thClass": "gl-w-10p",
  },
  Object {
    "end": 2019-12-31T23:59:59.000Z,
    "key": "3-months-ago",
    "label": "Dec",
    "start": 2019-12-01T00:00:00.000Z,
    "thClass": "gl-w-10p",
  },
  Object {
    "end": 2020-01-31T23:59:59.000Z,
    "key": "2-months-ago",
    "label": "Jan",
    "start": 2020-01-01T00:00:00.000Z,
    "thClass": "gl-w-10p",
  },
  Object {
    "end": 2020-02-29T23:59:59.000Z,
    "key": "1-months-ago",
    "label": "Feb",
    "start": 2020-02-01T00:00:00.000Z,
    "thClass": "gl-w-10p",
  },
  Object {
    "end": 2020-03-31T00:00:00.000Z,
    "key": "this-month",
    "label": "Mar",
    "start": 2020-03-01T00:00:00.000Z,
    "thClass": "gl-w-10p",
  },
]
`;

exports[`AI impact Dashboard utils generateDateRanges returns the expected date ranges on the first of the month 1`] = `
Array [
  Object {
    "end": 2021-02-28T23:59:59.000Z,
    "key": "5-months-ago",
    "label": "Feb",
    "start": 2021-02-01T00:00:00.000Z,
    "thClass": "gl-w-10p",
  },
  Object {
    "end": 2021-03-31T23:59:59.000Z,
    "key": "4-months-ago",
    "label": "Mar",
    "start": 2021-03-01T00:00:00.000Z,
    "thClass": "gl-w-10p",
  },
  Object {
    "end": 2021-04-30T23:59:59.000Z,
    "key": "3-months-ago",
    "label": "Apr",
    "start": 2021-04-01T00:00:00.000Z,
    "thClass": "gl-w-10p",
  },
  Object {
    "end": 2021-05-31T23:59:59.000Z,
    "key": "2-months-ago",
    "label": "May",
    "start": 2021-05-01T00:00:00.000Z,
    "thClass": "gl-w-10p",
  },
  Object {
    "end": 2021-06-30T23:59:59.000Z,
    "key": "1-months-ago",
    "label": "Jun",
    "start": 2021-06-01T00:00:00.000Z,
    "thClass": "gl-w-10p",
  },
  Object {
    "end": 2021-07-01T00:00:00.000Z,
    "key": "this-month",
    "label": "Jul",
    "start": 2021-07-01T00:00:00.000Z,
    "thClass": "gl-w-10p",
  },
]
`;

exports[`AI impact Dashboard utils generateDateRanges returns the expected date ranges on the last of the month 1`] = `
Array [
  Object {
    "end": 2020-10-31T23:59:59.000Z,
    "key": "5-months-ago",
    "label": "Oct",
    "start": 2020-10-01T00:00:00.000Z,
    "thClass": "gl-w-10p",
  },
  Object {
    "end": 2020-11-30T23:59:59.000Z,
    "key": "4-months-ago",
    "label": "Nov",
    "start": 2020-11-01T00:00:00.000Z,
    "thClass": "gl-w-10p",
  },
  Object {
    "end": 2020-12-31T23:59:59.000Z,
    "key": "3-months-ago",
    "label": "Dec",
    "start": 2020-12-01T00:00:00.000Z,
    "thClass": "gl-w-10p",
  },
  Object {
    "end": 2021-01-31T23:59:59.000Z,
    "key": "2-months-ago",
    "label": "Jan",
    "start": 2021-01-01T00:00:00.000Z,
    "thClass": "gl-w-10p",
  },
  Object {
    "end": 2021-02-28T23:59:59.000Z,
    "key": "1-months-ago",
    "label": "Feb",
    "start": 2021-02-01T00:00:00.000Z,
    "thClass": "gl-w-10p",
  },
  Object {
    "end": 2021-03-31T00:00:00.000Z,
    "key": "this-month",
    "label": "Mar",
    "start": 2021-03-01T00:00:00.000Z,
    "thClass": "gl-w-10p",
  },
]
`;

exports[`AI impact Dashboard utils generateSkeletonTableData returns the skeleton based on the table fields 1`] = `
Array [
  Object {
    "invertTrendColor": true,
    "metric": Object {
      "identifier": "cycle_time",
      "value": "Cycle time",
    },
  },
  Object {
    "invertTrendColor": true,
    "metric": Object {
      "identifier": "lead_time",
      "value": "Lead time",
    },
  },
  Object {
    "invertTrendColor": undefined,
    "metric": Object {
      "identifier": "deployment_frequency",
      "value": "Deployment frequency",
    },
  },
  Object {
    "invertTrendColor": true,
    "metric": Object {
      "identifier": "change_failure_rate",
      "value": "Change failure rate",
    },
  },
  Object {
    "invertTrendColor": true,
    "metric": Object {
      "identifier": "vulnerability_critical",
      "value": "Critical vulnerabilities over time",
    },
  },
  Object {
    "invertTrendColor": undefined,
    "metric": Object {
      "identifier": "code_suggestions_usage_rate",
      "value": "Code Suggestions usage",
    },
  },
]
`;

exports[`AI impact Dashboard utils generateTableColumns returns the expected table fields in a leap year 1`] = `
Array [
  Object {
    "key": "metric",
    "label": "Metric",
    "thClass": "gl-w-20p",
  },
  Object {
    "end": 2019-10-31T23:59:59.000Z,
    "key": "5-months-ago",
    "label": "Oct",
    "start": 2019-10-01T00:00:00.000Z,
    "thClass": "gl-w-10p",
  },
  Object {
    "end": 2019-11-30T23:59:59.000Z,
    "key": "4-months-ago",
    "label": "Nov",
    "start": 2019-11-01T00:00:00.000Z,
    "thClass": "gl-w-10p",
  },
  Object {
    "end": 2019-12-31T23:59:59.000Z,
    "key": "3-months-ago",
    "label": "Dec",
    "start": 2019-12-01T00:00:00.000Z,
    "thClass": "gl-w-10p",
  },
  Object {
    "end": 2020-01-31T23:59:59.000Z,
    "key": "2-months-ago",
    "label": "Jan",
    "start": 2020-01-01T00:00:00.000Z,
    "thClass": "gl-w-10p",
  },
  Object {
    "end": 2020-02-29T23:59:59.000Z,
    "key": "1-months-ago",
    "label": "Feb",
    "start": 2020-02-01T00:00:00.000Z,
    "thClass": "gl-w-10p",
  },
  Object {
    "end": 2020-03-31T00:00:00.000Z,
    "key": "this-month",
    "label": "Mar",
    "start": 2020-03-01T00:00:00.000Z,
    "thClass": "gl-w-10p",
  },
  Object {
    "description": "Past 6 Months",
    "end": 2020-02-29T23:59:59.000Z,
    "key": "change",
    "label": "Change (%)",
    "start": 2019-09-01T00:00:00.000Z,
    "thClass": "gl-w-20p",
  },
]
`;

exports[`AI impact Dashboard utils generateTableColumns returns the expected table fields on the first of the month 1`] = `
Array [
  Object {
    "key": "metric",
    "label": "Metric",
    "thClass": "gl-w-20p",
  },
  Object {
    "end": 2021-02-28T23:59:59.000Z,
    "key": "5-months-ago",
    "label": "Feb",
    "start": 2021-02-01T00:00:00.000Z,
    "thClass": "gl-w-10p",
  },
  Object {
    "end": 2021-03-31T23:59:59.000Z,
    "key": "4-months-ago",
    "label": "Mar",
    "start": 2021-03-01T00:00:00.000Z,
    "thClass": "gl-w-10p",
  },
  Object {
    "end": 2021-04-30T23:59:59.000Z,
    "key": "3-months-ago",
    "label": "Apr",
    "start": 2021-04-01T00:00:00.000Z,
    "thClass": "gl-w-10p",
  },
  Object {
    "end": 2021-05-31T23:59:59.000Z,
    "key": "2-months-ago",
    "label": "May",
    "start": 2021-05-01T00:00:00.000Z,
    "thClass": "gl-w-10p",
  },
  Object {
    "end": 2021-06-30T23:59:59.000Z,
    "key": "1-months-ago",
    "label": "Jun",
    "start": 2021-06-01T00:00:00.000Z,
    "thClass": "gl-w-10p",
  },
  Object {
    "end": 2021-07-01T00:00:00.000Z,
    "key": "this-month",
    "label": "Jul",
    "start": 2021-07-01T00:00:00.000Z,
    "thClass": "gl-w-10p",
  },
  Object {
    "description": "Past 6 Months",
    "end": 2021-06-30T23:59:59.000Z,
    "key": "change",
    "label": "Change (%)",
    "start": 2021-01-01T00:00:00.000Z,
    "thClass": "gl-w-20p",
  },
]
`;

exports[`AI impact Dashboard utils generateTableColumns returns the expected table fields on the last of the month 1`] = `
Array [
  Object {
    "key": "metric",
    "label": "Metric",
    "thClass": "gl-w-20p",
  },
  Object {
    "end": 2020-10-31T23:59:59.000Z,
    "key": "5-months-ago",
    "label": "Oct",
    "start": 2020-10-01T00:00:00.000Z,
    "thClass": "gl-w-10p",
  },
  Object {
    "end": 2020-11-30T23:59:59.000Z,
    "key": "4-months-ago",
    "label": "Nov",
    "start": 2020-11-01T00:00:00.000Z,
    "thClass": "gl-w-10p",
  },
  Object {
    "end": 2020-12-31T23:59:59.000Z,
    "key": "3-months-ago",
    "label": "Dec",
    "start": 2020-12-01T00:00:00.000Z,
    "thClass": "gl-w-10p",
  },
  Object {
    "end": 2021-01-31T23:59:59.000Z,
    "key": "2-months-ago",
    "label": "Jan",
    "start": 2021-01-01T00:00:00.000Z,
    "thClass": "gl-w-10p",
  },
  Object {
    "end": 2021-02-28T23:59:59.000Z,
    "key": "1-months-ago",
    "label": "Feb",
    "start": 2021-02-01T00:00:00.000Z,
    "thClass": "gl-w-10p",
  },
  Object {
    "end": 2021-03-31T00:00:00.000Z,
    "key": "this-month",
    "label": "Mar",
    "start": 2021-03-01T00:00:00.000Z,
    "thClass": "gl-w-10p",
  },
  Object {
    "description": "Past 6 Months",
    "end": 2021-02-28T23:59:59.000Z,
    "key": "change",
    "label": "Change (%)",
    "start": 2020-09-01T00:00:00.000Z,
    "thClass": "gl-w-20p",
  },
]
`;

exports[`AI impact Dashboard utils generateTableRows returns the data formatted as a table row 1`] = `
Object {
  "change_failure_rate": Object {
    "1-months-ago": Object {
      "value": "7.5%",
    },
    "2-months-ago": Object {
      "value": "0.0%",
    },
    "3-months-ago": Object {
      "value": "0.0%",
    },
    "4-months-ago": Object {
      "value": "5.6%",
    },
    "5-months-ago": Object {
      "value": "8.9%",
    },
    "change": Object {
      "value": -0.15730337078651688,
    },
    "this-month": Object {
      "value": "4.0%",
    },
  },
  "deployment_frequency": Object {
    "1-months-ago": Object {
      "value": "-",
    },
    "2-months-ago": Object {
      "value": "30.0/d",
    },
    "3-months-ago": Object {
      "value": "0.0/d",
    },
    "4-months-ago": Object {
      "value": "15.0/d",
    },
    "5-months-ago": Object {
      "value": "10.0/d",
    },
    "change": Object {
      "value": 0,
    },
    "this-month": Object {
      "value": "30.0/d",
    },
  },
}
`;
